In [56]:
import pandas as pd
import numpy as np
import os

# Define the project folder path
project_folder = r"/Users/cem/Desktop/Data Immersion/Achievement 4_Project"

# Define the data folder path
data_folder = os.path.join(project_folder, "Data", "4.3_orders_products")

# Load orders.csv
orders_path = os.path.join(data_folder, "orders.csv")
df_ords = pd.read_csv(orders_path, index_col=False)

# Display the first 5 rows
df_ords.head()

order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0   2539329        1    prior             1          2                  8   
1   2398795        1    prior             2          3                  7   
2    473747        1    prior             3          3                 12   
3   2254736        1    prior             4          4                  7   
4    431534        1    prior             5          4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [57]:
# Load products.csv
products_path = os.path.join(data_folder, "products.csv")
df_prods = pd.read_csv(products_path, index_col=False)

# Display the first 5 rows
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [58]:
# Load departments.csv
departments_path = os.path.join(data_folder, "departments.csv")
df_dep = pd.read_csv(departments_path, index_col=False)

# Display the first 5 rows
df_dep.head()

department_id       1      2       3        4        5              6  \
0    department  frozen  other  bakery  produce  alcohol  international   

           7     8                9  ...            12      13         14  \
0  beverages  pets  dry goods pasta  ...  meat seafood  pantry  breakfast   

             15          16         17      18      19    20       21  
0  canned goods  dairy eggs  household  babies  snacks  deli  missing  

[1 rows x 22 columns]

In [61]:
# Q2
print(df_ords.columns)

Index(['order_id', 'user_id', 'eval_set', 'order_number', 'order_dow',
       'order_hour_of_day', 'days_since_prior_order'],
      dtype='object')


In [62]:
# Q2: Change order_number to a string
df_ords['order_number'] = df_ords['order_number'].astype(str)

# Check data types
print(df_ords.dtypes)

order_id                    int64
user_id                     int64
eval_set                   object
order_number               object
order_dow                   int64
order_hour_of_day           int64
days_since_prior_order    float64
dtype: object


In [63]:
# Q3: Rename 'order_dow' to 'order_day_of_week' without overwriting the original DataFrame
df_ords_renamed = df_ords.rename(columns={'order_dow': 'order_day_of_week'})

# Display the first 5 rows of the renamed DataFrame
df_ords_renamed.head()


order_id  user_id eval_set order_number  order_day_of_week  \
0   2539329        1    prior            1                  2   
1   2398795        1    prior            2                  3   
2    473747        1    prior            3                  3   
3   2254736        1    prior            4                  4   
4    431534        1    prior            5                  4   

   order_hour_of_day  days_since_prior_order  
0                  8                     NaN  
1                  7                    15.0  
2                 12                    21.0  
3                  7                    29.0  
4                 15                    28.0

In [64]:
# Q4: Calculate frequency of orders by hour
order_hour_frequency = df_ords['order_hour_of_day'].value_counts().sort_index()

# Find the busiest hour
busiest_hour = order_hour_frequency.idxmax()
order_count = order_hour_frequency.max()

# Display the results
print("Frequency of orders by hour:")
print(order_hour_frequency)
print(f"\nThe busiest hour for placing orders is {busiest_hour} with {order_count} orders.")


Frequency of orders by hour:
order_hour_of_day
0      22758
1      12398
2       7539
3       5474
4       5527
5       9569
6      30529
7      91868
8     178201
9     257812
10    288418
11    284728
12    272841
13    277999
14    283042
15    283639
16    272553
17    228795
18    182912
19    140569
20    104292
21     78109
22     61468
23     40043
Name: count, dtype: int64

The busiest hour for placing orders is 10 with 288418 orders.


In [71]:
# Q5
# Transpose
df_dep_t = df_dep.T

# Use first row as header
new_header = df_dep_t.iloc[0]

# Create new DataFrame from row 1 onward
df_dep_t_new = df_dep_t[1:]

# Set new header row as columns
df_dep_t_new.columns = new_header

# Display transformed departments DataFrame
df_dep_t_new.head()

# The header is now fixed and the rows-columns reversed with transpose:

department_id department
1                 frozen
2                  other
3                 bakery
4                produce
5                alcohol

In [69]:
## Creating data dictionary
data_dict = df_dep_t_new.to_dict('index')

data_dict

# 4 means to the category "produce".

None


{0: {'department': 'frozen'},
 1: {'department': 'other'},
 2: {'department': 'bakery'},
 3: {'department': 'produce'},
 4: {'department': 'alcohol'},
 5: {'department': 'international'},
 6: {'department': 'beverages'},
 7: {'department': 'pets'},
 8: {'department': 'dry goods pasta'},
 9: {'department': 'bulk'},
 10: {'department': 'personal care'},
 11: {'department': 'meat seafood'},
 12: {'department': 'pantry'},
 13: {'department': 'breakfast'},
 14: {'department': 'canned goods'},
 15: {'department': 'dairy eggs'},
 16: {'department': 'household'},
 17: {'department': 'babies'},
 18: {'department': 'snacks'},
 19: {'department': 'deli'},
 20: {'department': 'missing'}}

In [73]:
# Q6
#creating the subset
df_prods_breakfast = df_prods.loc[df_prods['department_id'] == 14]

df_prods_breakfast.head()

product_id                                      product_name  aisle_id  \
27           28                                 Wheat Chex Cereal       121   
33           34                                               NaN       121   
67           68                           Pancake Mix, Buttermilk       130   
89           90                                      Smorz Cereal       121   
210         211  Gluten Free Organic Cereal Coconut Maple Vanilla       130   

     department_id  prices  
27              14    10.1  
33              14    12.2  
67              14    13.7  
89              14     3.9  
210             14     3.6

In [74]:
# Q7: Dinner Party Subset
df_prods_dinner_party = df_prods.loc[df_prods['department_id'].isin([5, 20, 7, 12])]

df_prods_dinner_party.head()


product_id                                    product_name  aisle_id  \
2            3            Robust Golden Unsweetened Oolong Tea        94   
6            7                  Pure Coconut Water With Orange        98   
9           10  Sparkling Orange Juice & Prickly Pear Beverage       115   
10          11                               Peach Mango Juice        31   
16          17                               Rendered Duck Fat        35   

    department_id  prices  
2               7     4.5  
6               7     4.4  
9               7     8.4  
10              7     2.8  
16             12    17.1

In [75]:
# Q8: Number of rows in dinner party subset
df_prods_dinner_party.shape


(7650, 5)

In [82]:
#Step 9: finding info about customer with user_id 1
df_user1 = df_ords.loc[df_ords['user_id'] == 1]

df_user1

order_id  user_id eval_set order_number  order_dow  order_hour_of_day  \
0    2539329        1    prior            1          2                  8   
1    2398795        1    prior            2          3                  7   
2     473747        1    prior            3          3                 12   
3    2254736        1    prior            4          4                  7   
4     431534        1    prior            5          4                 15   
5    3367565        1    prior            6          2                  7   
6     550135        1    prior            7          1                  9   
7    3108588        1    prior            8          1                 14   
8    2295261        1    prior            9          1                 16   
9    2550362        1    prior           10          4                  8   
10   1187899        1    train           11          4                  8   

    days_since_prior_order  
0                      NaN  
1                     15.0  
2                     21.0  
3                     29.0  
4                     28.0  
5                     19.0  
6                     20.0  
7                     14.0  
8                      0.0  
9                     30.0  
10                    14.0

In [84]:
# Other stats of the user's with id=1
df_user1.describe()

order_id  user_id  order_dow  order_hour_of_day  \
count  1.100000e+01     11.0  11.000000          11.000000   
mean   1.923450e+06      1.0   2.636364          10.090909   
std    1.071950e+06      0.0   1.286291           3.477198   
min    4.315340e+05      1.0   1.000000           7.000000   
25%    8.690170e+05      1.0   1.500000           7.500000   
50%    2.295261e+06      1.0   3.000000           8.000000   
75%    2.544846e+06      1.0   4.000000          13.000000   
max    3.367565e+06      1.0   4.000000          16.000000   

       days_since_prior_order  
count               10.000000  
mean                19.000000  
std                  9.030811  
min                  0.000000  
25%                 14.250000  
50%                 19.500000  
75%                 26.250000  
max                 30.000000

In [85]:
# Q12: Export df_ords to 'Prepared Data' folder
prepared_data_folder = os.path.join(project_folder, 'Data', 'Prepared Data')
os.makedirs(prepared_data_folder, exist_ok=True)

df_ords.to_csv(os.path.join(prepared_data_folder, 'orders_wrangled.csv'), index=False)


In [86]:
# Q13: Export df_dep_t_new to 'Prepared Data' folder
df_dep_t_new.to_csv(os.path.join(prepared_data_folder, 'departments_wrangled.csv'), index=False)


# Insights for Final Report

### Insight 1: Orders Data Wrangling

- We cleaned the orders data by dropping the unnecessary 'eval_set' column and converting identifier variables like 'order_id', 'user_id', and 'order_number' to strings.
- This step ensures that these columns are treated as categorical variables rather than numerical, preventing unintended calculations. The cleaned dataset is now easier to analyze and aligns with best practices for handling identifier variables.


### Insight 2: Department Data Wrangling

- We transposed and reformatted the departments dataset so that 'department_id' and 'department' are clearly defined as separate columns. This restructuring allowed us to easily match products to their corresponding departments using consistent identifiers.
- The cleaned dataset was exported as a CSV file for future use.


### Insight 3: Busiest Hour for Orders

- The analysis revealed that the busiest hour for placing orders is 10am, with over 288,000 orders placed during that time slot.
- This insight is valuable for the sales and marketing team, who can target promotions during this peak period. Additionally, understanding hourly ordering patterns helps optimize staffing and operational planning.


### Insight 4: Breakfast Department

- We identified the breakfast department using 'department_id' 14, which allows for targeted analysis of breakfast product sales.
- This subset can be further analyzed to determine which breakfast items are most popular and whether specific promotions could boost sales.
- Knowing the exact department_id simplifies future filtering and reporting.


### Insight 5: Dinner Party Products

- A subset of dinner party-related products was created by filtering for department_ids 5 (alcohol), 7 (beverages), 12 (meat/seafood), and 20 (deli).
- This subset enables targeted marketing analysis to identify trends and plan promotions around these product categories.
- The dataset contains 7,650 rows, representing a substantial portion of the catalog relevant to dinner events.


### Insight 6: User ID 1 Analysis

- User ID 1 exhibits specific ordering behavior, typically placing orders early in the week and in the early hours of the day. The analysis shows that this user averages approximately every 19 days between orders.
- These insights could be leveraged by the marketing team for personalized promotions or loyalty program offers.


### Insight 7: Data Exports

- Both the cleaned orders dataset and the departments dataset were exported as CSV files.
- This step ensures that the cleaned and structured data is available for additional analysis, reporting, and visualization, saving time for future tasks.
- This practice aligns with project deliverables and documentation best practices.